src/runtime/refiner.ts

```diff
@@ -1237,143 +1241,220 @@ export class Fraction {
+    for (const tKey of Object.keys(this.terms)) {
+      const term = Term.fromKey(tKey);
+      for (const indeterminate of Object.keys(term.indeterminates)) {
+        indeterminates.add(indeterminate);
+      }
     }
-    if (pow === 1) {
-      return new FieldNamePrimitive(fn, Primitive.NUMBER);
+    return indeterminates;
+  }
+
+  isUnivariate(): boolean {
+    return this.getIndeterminates().size === 1;
+  }
+
+  degree(): number {
+    let degree = 0;
+    for (const tKey of Object.keys(this.terms)) {
+      const term = Term.fromKey(tKey);
+      let sum = 0;
+      for (const power of Object.values(term.indeterminates)) {
+        sum += power;
+      }
+      degree = sum > degree ? sum : degree;
     }
-    return new BinaryExpression(
-      Polynomial.inderminateToExpression(1, fn),
-      Polynomial.inderminateToExpression(pow-1, fn),
-      new RefinementOperator(Op.MUL));
+    return degree;
   }
 
-  // returns ax^n + bx^n-1 + ... c  <op> 0
+  // returns <multinomial> <op> CONSTANT
   toExpression(op: Op): RefinementExpression {
-    if (this.degree() === 0) {
+    if (this.isConstant()) {
       return new BinaryExpression(
-        new NumberPrimitive(this.coeffs[0]),
+        new NumberPrimitive(this.isZero() ? 0 : this.terms[CONSTANT]),
         new NumberPrimitive(0),
         new RefinementOperator(op));
     }
-    if (!this.indeterminate) {
-      throw new Error('FieldName not found!');
-    }
-    if (this.degree() === 1) {
-      // ax+b <op1> 0 => x <op2> -b/a
+    if (this.isUnivariate() && this.degree() === 1) {
       const operator = new RefinementOperator(op);
-      if (this.coeffs[1] < 0) {
+      const indeterminate = this.getIndeterminates().values().next().value;
+      // TODO(ragdev): Implement a neater way to get the leading coefficient
+      const leadingCoeff = this.terms[`{"${indeterminate}":1}`];
+      const cnst = this.terms.hasOwnProperty(CONSTANT) ? this.terms[CONSTANT] : 0;
+      if (leadingCoeff < 0) {
         operator.flip();
       }
       return new BinaryExpression(
-        new FieldNamePrimitive(this.indeterminate, Primitive.NUMBER),
-        new NumberPrimitive(-this.coeffs[0]/this.coeffs[1]),
+        new FieldNamePrimitive(indeterminate, Primitive.NUMBER),
+        new NumberPrimitive(-cnst/leadingCoeff),
         operator);
     }
     let expr = null;
-    for (let i = this.coeffs.length - 1; i >= 0; i -= 1) {
-      if (this.coeffs[i] === 0) {
-        continue;
-      }
-      let termExpr = null;
-      if (i > 0) {
-        termExpr = Polynomial.inderminateToExpression(i, this.indeterminate);
-        if (Math.abs(this.coeffs[i]) !== 1) {
+    let cnst = 0;
+    for (const [tKey, tCoeff] of Object.entries(this.terms)) {
+      if (tKey === CONSTANT) {
+        cnst = tCoeff;
+      } else {
+        const term = Term.fromKey(tKey);
+        let termExpr = term.toExpression();
+        if (tCoeff !== 1) {
           termExpr = new BinaryExpression(
```

Comment:
This is getting pretty complicated: Maybe it should be a private helper function and have a name so we can more easily see what it is doing?

Reply:
I see. I'll try to refactor this.